nexus/deployment/src/blueprint_builder.rs

```diff
@@ -329,6 +455,118 @@ impl<'a> BlueprintBuilder<'a> {
+        // goes through the blueprint system (which seems likely), we'll need to
+        // check that we're if this builder is being used to make such a change,
+        // that change is also reflected here in a new zone. Perhaps these
+        // settings should be part of `Policy` instead?
+        let (external_tls, external_dns_servers) = self
+            .parent_blueprint
+            .omicron_zones
+            .values()
+            .find_map(|sled_zones| {
+                sled_zones.zones.iter().find_map(|z| match &z.zone_type {
+                    OmicronZoneType::Nexus {
+                        external_tls,
+                        external_dns_servers,
+                        ..
+                    } => Some((*external_tls, external_dns_servers.clone())),
+                    _ => None,
+                })
+            })
+            .ok_or(Error::NoNexusZonesInParentBlueprint)?;
+
+        let nexus_id = Uuid::new_v4();
+        let external_ip = self
+            .available_external_ips
+            .next()
+            .ok_or(Error::NoExternalServiceIpAvailable)?;
+
+        let nic = {
+            let (ip, subnet) = match external_ip {
+                IpAddr::V4(_) => (
+                    self.nexus_v4_ips
+                        .next()
+                        .ok_or(Error::ExhaustedNexusIps)?
+                        .into(),
+                    IpNet::from(*NEXUS_OPTE_IPV4_SUBNET).into(),
+                ),
+                IpAddr::V6(_) => (
+                    self.nexus_v6_ips
+                        .next()
+                        .ok_or(Error::ExhaustedNexusIps)?
+                        .into(),
+                    IpNet::from(*NEXUS_OPTE_IPV6_SUBNET).into(),
+                ),
+            };
+            NetworkInterface {
+                id: Uuid::new_v4(),
+                kind: NetworkInterfaceKind::Service(nexus_id),
+                name: format!("nexus-{nexus_id}").parse().unwrap(),
+                ip,
+                mac: self.random_mac(),
+                subnet,
+                vni: Vni::SERVICES_VNI,
+                primary: true,
+                slot: 0,
+            }
+        };
+
+        let ip = self.sled_alloc_ip(sled_id)?;
+        let port = omicron_common::address::NEXUS_INTERNAL_PORT;
+        let internal_address = SocketAddrV6::new(ip, port, 0, 0).to_string();
+        let zone = OmicronZoneConfig {
+            id: nexus_id,
+            underlay_address: ip,
+            zone_type: OmicronZoneType::Nexus {
+                internal_address,
+                external_ip,
+                nic,
+                external_tls,
+                external_dns_servers,
+            },
+        };
+        self.sled_add_zone(sled_id, zone)?;
+        Ok(Ensure::Added)
+    }
+
+    fn random_mac(&mut self) -> MacAddr {
+        let mut mac = MacAddr::random_system();
+
+        // TODO-performance if the size of `used_macs` starts to approach some
```

Comment:
Feel free to ignore this.  It's not a well fleshed-out thought.  I worry about using randomness in the planner because I think that'll make it harder to test, and easier to introduce thrashing behavior (where two Nexuses -- or even one! -- fight over an arbitrary decision).  You could as well argue the other way around, that by using randomness we'll be forced to encounter and deal with thrashing.  I guess it's just a gut feel that randomness in the planner is going to make our lives harder.

Reply:
Hmm! It's a fair concern. My gut feeling is that there are going to be other places where we want "just pick a random one" (e.g., which of the N sleds should get an extra Nexus/cockroach/etc.), and if we do something _too_ deterministic (e.g., sort by sled ID) we'll make our lives unnecessarily difficult by picking the same sled for too many things.

Maybe a bad idea: what if the planner seeded an RNG with something deterministic, maybe the parent blueprint ID? Then it has deterministic behavior for any single given parent, but we can still freely use randomness.

Reply:
> Maybe a bad idea: what if the planner seeded an RNG with something deterministic, maybe the parent blueprint ID? Then it has deterministic behavior for any single given parent, but we can still freely use randomness.

I am a fan of this! It's basically how proptest and all the other related tools give you replay determinism.

Reply:
> we'll make our lives unnecessarily difficult by picking the same sled for too many things

I don't think randomness really solves that problem, but admittedly it might be good enough for a while.  I think I'd propose an approach where we narrow the set of sleds to which ones are acceptable for a choice, then pick deterministically among those.  That deterministic process could initially be "pick the lexicographically-first sled id", but it could also be "pick the one with fewer zones on it" or "pick the one with the least CPU/disk/memory already allocated by the control plane" or "assign each zone type a score based on how much load we expect it to generate and then pick the sled with the lowest total score".  That is, we can do something simple (pick the first) or sophisticated (if we find we need to deal better with distribution) but it's still deterministic.

This is a lot more complicated than generating MAC addresses, though.  Again, feel free to ignore this but I mention it because the sequential allocator seems no worse here and eliminates the only current source of randomness in the planner (I think).

> Maybe a bad idea: what if the planner seeded an RNG with something deterministic, maybe the parent blueprint ID? Then it has deterministic behavior for any single given parent, but we can still freely use randomness.

I'm not sure.  I could see that going either way too.  It doesn't help test determinism unless we ensure that we use deterministic UUID generation too.  I looked into that briefly while writing tests but it seemed like it'd require a lot of surgery.

Reply:
> Maybe a bad idea: what if the planner seeded an RNG with something deterministic, maybe the parent blueprint ID? Then it has deterministic behavior for any single given parent, but we can still freely use randomness.

I think a deterministic source of randomness stored in the blueprint would be cool, and would probably end up getting used for a lot of things.

> It doesn't help test determinism unless we ensure that we use deterministic UUID generation too. I looked into that briefly while writing tests but it seemed like it'd require a lot of surgery.

Deterministic UUID generation would be pretty useful as well I think, effectively requiring that a source of UUIDs be passed down.

Reply:
> This is a lot more complicated than generating MAC addresses, though. Again, feel free to ignore this but I mention it because the sequential allocator seems no worse here and eliminates the only current source of randomness in the planner (I think).

Oh, /facepalm, MACs are not the only source of random output of blueprints, specifically because of UUIDs (which you and @sunshowers mentioned - thank you). We already generate the NTP and Crucible zone IDs, and this PR adds two more UUID generations: the Nexus zone ID and the ID for the Nexus zone NIC.

I think it would not be a huge lift to add deterministic random to `BlueprintBuilder` and use it for both UUIDs and MACs. Probably the trickiest bit will be ensuring future changes continue to use the deterministic random, but probably we can catch that in unit tests? I'll give it a whirl and see what it looks like.

Reply:
Yeah, thinking about it, it's a bit hard to enforce the invariant that people don't use the `Uuid::new_v4` method long-term. One way is probably to not enable the `v4` feature, but that seems excessive (using the system RNG is completely fine in most other spots, just not here).

I guess a unit test could generate blueprints with a fixed seed several times, and ensure that the generated UUIDs stay the same.

Reply:
Another idea really is to just ban `new_v4` via clippy: https://rust-lang.github.io/rust-clippy/master/#/disallowed_methods

Reply:
I think it's fine if we want to have a deterministic source of randomness for blueprints.  At the same time, I think randomness is not a great solution to any of the problems we've mentioned, or at least not better than an alternative (like sequential MAC address allocation), so I'm not sure it's worth the effort.

I think my aversion to randomness here comes down to a few things:

- For automated testing, it's more annoying to verify that the planner generated the blueprint we expected because we can't just hardcode the values we know it will choose. Deterministic-random doesn't _necessarily_ help here but it could.  But we already wrote a higher-level `diff` thing that can tell if two blueprints are different, ignoring things like the "id" and "time_created" (which will always be different anyway).  So this isn't that big a deal.
- When we get to making this more autonomous, it will become important for Nexus instances to be able to look at some other instance's plan and say "is that okay?"  That _might_ be the same as "is that equivalent to the thing I would do?" -- or it might not.  I alluded to this earlier.  Choosing deterministically makes it easier for us to punt on this for longer (because two Nexus instances will tend to create the same plans), but it doesn't really solve the problem.  Also: we're talking about hypotheticals on hypotheticals at this point and I really don't think we want to spend a lot of time on such things now.
- Probably the biggest thing: during general operation and especially in support contexts: it means the system's behavior is less predictable.  I can imagine wanting to set a customer's expectations -- or even our own -- about what would happen when the planner is enabled or when given some input.  With randomness, even deterministic randomness, that's harder.
- I've often reached for random approaches in the past and found that they work okay for a while, but in the limit / at scale, the variance winds up being a problem.

Honestly, none of these is that compelling.  If there were a compelling reason to use randomness here, I'd say go for it.  But I don't see any reason _to_ use randomness, either, aside from being slightly quicker to implement than sequential.  Either way, I don't think it's worth investing a lot of time on this part right now.

Reply:
I went through this and basically copied what RSS was doing, so I ended up with sequential OPTE IP address generation and random MAC generation. It's a small change to switch over to sequential MAC generation; I can put that in this PR so I'm not introducing a new kind of randomness, and we can punt on whether we should have deterministic UUIDs for a followup.

Reply:
[9748940](https://github.com/oxidecomputer/omicron/pull/4959/commits/97489400c6221be65d165ee400ad264dc9dfbda7) switches to sequential MAC allocation. I'm not going to mark this conversation resolved because I don't want to hide it, but I think it's resolved as far as this PR is concerned; please correct me if that's wrong!